### ISSUE TITLE
File is not writable error

### ISSUE THREAD

**Issue by Venomz3** [o10n-x/wordpress-javascript-optimization#15]
I tryed to set async JavaScript and set config filter and I got this error when saving setting.
Error during proxy update (shutdown callback): 
Failed to proxy https://static.xx.fbcdn.net/rsrc.php/v3/y-/r/MC9xhleHEfF.js Error: Failed to store cache file /wp-content/cache/o10n/js/proxy/87/18/12/b3264aebead032ea2ad474ad2e.js
File is not writeable: /wp-content/cache/o10n/js/proxy/87/18/12/b3264aebead032ea2ad474ad2e.js

Failed to store cache file /wp-content/cache/o10n/js/proxy/87/18/12/b3264aebead032ea2ad474ad2e.js
File is not writeable: /wp-content/cache/o10n/js/proxy/87/18/12/b3264aebead032ea2ad474ad2e.js

Config filter: "[
  {
    "match": "/https://www.jhhypnoosi.fi/wp-admin/admin-ajax.php?lang.en&action.bookly_render_service&csrf_token.+/",
    "regex": true,
    "async": false,
    "rel_preload": false,
    "exec_timing": {
      "type": "requestAnimationFrame",
      "frame": 1
    },
    "localStorage": {
      "max_size": 10000,
      "update_interval": 3600,
      "expire": 86400,
      "head_update": true
    }
  }
]"

I`m waiting answer from my server if this is also related to my server.  I asked 2 questions. So hopefully they answer be tomorrow.  
Additional info (maybe unrelated but not sure) I used Bulletproof security plugin`s wizard to make sure that all plugins are whitelisted. After that my php version was downgraded to 5.6. Immediately I changed php back to 7.0. At time when I updated all o10n plugins  (today) php version was 5.6 then I realized that php version was changed. Now I know that bulletproof security can change php version also i can avoid that in future.    Earlier  when I got errors I did not do anything with bulletproof security.

**Comment by optimalisatie** [comment]
This is the result of the latest update that intends to ignore the issue on your server in which `file_exists` returns false while `mkdir` returns `mkdir(): File exists`. 

Even after using `clearstatcache` and double checking using `file_exists` and `is_dir` the issue persists, indicating that it may be a file system related bug specific to your server or an other bug (potentially related to PHP shutdown and `clearstatcache`).

Apparently, as shown by the error that you reported, the directory does not exist causing the logic that requires the directory to fail.

It would make sense to rely on `file_exists` and consider the `mkdir` error to be invalid. 

As the error occurs in a PHP shutdown callback, I am suspecting that it could be related to a not functioning `clearstatcache` issue in which mkdir does not have access to the latest file system info but instead relies on a cache from before the shutdown. What may be happening is that you empty the cache before PHP shutdown and that PHP continues to execute methods that rely on cache directories in a shutdown callback, directories that apparently maintain a 'exists' state despite that they do not exist on the disk and despite the use of `clearstatcache`.

http://php.net/manual/en/function.clearstatcache.php

We will change the `mkdir` method in the next update to throw an error when the issue occurs.

**Comment by Venomz3** [comment]
Server support contacted me back and wanted to investigate problems and asked access to admin panel. They wanted to see your code. Or otherwords code that causes problem. 
So it might take few days as it`s weekend now. 

Also I contacted WMPL plugin`s support because Pingdom tools performance site show abnormal things on page loading. 
It shows that some files loaded twice for eg. when testing Finnish lanuage version then english file is loaded somewere in the list or couple of files. Then I disable all wpo plugins and to see if there is some inteference. It`s like masking problem when css and js files are concatated.
When I disable wpo plugins and cleared caches and did several times pingdon test to that same page there are about 120 file requests. Files are requested doulble times now when I disables wpo plugins.  I suspect WPML 4.0.x has done someting or there is bug in pingdom test.. WPMP 4 version released 6.6. and I waited 2 minor versions to before I updated just to get bugs out. 

So I suspect that WPML (multilingual support for wordpess) could have something to do with performance issues or bulletproof security-plugin.

**Comment by optimalisatie** [comment]
All code is available on Github.

You can send them the following URL for the specific PHP logic that is failing:

https://github.com/o10n-x/wordpress-o10n-core/blob/master/controllers/file.class.php#L85-L165

**Comment by Venomz3** [comment]
I sent link to server support and I haven heard them yet about that issue. I got automated email that they are doing maintenance at weekend. So let`s hope that they clear issue. 
Just minute ago I got error messages (Haven`t had them a while): 
Error during proxy update (shutdown callback): 
Failed to proxy https://static.xx.fbcdn.net/rsrc.php/v3/yK/r/yiKiTmkQ69J.js Error: Failed to store cache file /wp-content/cache/o10n/js/proxy/59/c1/8a/34c99032ed8a30c05feb83e14d.js
File is not writeable: /wp-content/cache/o10n/js/proxy/59/c1/8a/34c99032ed8a30c05feb83e14d.js

Failed to store cache file /wp-content/cache/o10n/js/proxy/59/c1/8a/34c99032ed8a30c05feb83e14d.js
File is not writeable: /wp-content/cache/o10n/js/proxy/59/c1/8a/34c99032ed8a30c05feb83e14d.js
Failed to store cache file /wp-content/cache/o10n/css/proxy/12/5c/16/f039cd8194c59b4ca06d17fbeb.css
File is not writeable: /wp-content/cache/o10n/css/proxy/12/5c/16/f039cd8194c59b4ca06d17fbeb.css

All I did was clicked top menu to JavaScript optimization. I haven`t done anything for hours in wp-admin and then when I clicked that menu those errors came. Just a thought that came into my mind : " Like server was in sleep mode and when it wakes then it throws those error"  That has been a common thing few last time that I haven`t done nothing in wp-admin for few hours and when goto WPO then I got error.  Still it`s not that many hours that I have to relogin.

Another thing to  ask:  JavaScript optimization > code optimization > Minify exclude list
  I`ve tested to type filename only and path+filename and I`m not succeeding. Or is there a bug ? Those files are not excluded. If there is a bug I`ll report it as separate bug if I haven`t typed wrongly. 
How should exclude files should be typed in that box ??

**Comment by AlphaDE** [comment]
Without having changed the config, I see similar errors on Google, where Google crawler could not access cache files

/wp-content/cache/o10n/js/src/7f/5c/55/dc9947e51073136388668117d2.js.meta

Mostly these .meta files.

**Comment by optimalisatie** [comment]
There is no sleep mode, the errors are not real time. The error messages are simply queued until a optimization admin page is viewed.

The plugin uses modern error handling based on PHP Exceptions. This means that if anything within the plugin or PHP fails, it will be reported in the admin (just on the optimization admin pages, not global).

#### PHP Exceptions

A unique feature of the WPO plugins is that it uses PHP Exceptions that are logged to the WPO plugin admin so that you have a clear view if anything goes wrong. It also enables a website to remain usable if the WPO plugin (or parts of it) fail.

> ### [Why WP_Error Sucks](https://journal.rmccue.io/165/why-wp_error-sucks/)
> As anyone who has worked with WordPress knows, WordPress supported PHP 4 for a long time, even after many other projects had switched. The advantage of this was supporting significantly more hosts, with most PHP 5-only features either not being needed or being easy to reimplement.
> 
> One of the new features added to PHP in PHP 5 was exception handling... 
> 
> https://journal.rmccue.io/165/why-wp_error-sucks/

#### PHP errors in browser console

A unique innovation is that PHP errors are also displayed in the browser console when debug mode is enabled. Some of the Exceptions may not be fatal (e.g. a failed regular expression). With debug mode enabled you can find these errors in the browser console for optimization purposes.

![image](https://user-images.githubusercontent.com/8843669/41180460-a03ca1a4-6b6e-11e8-8cbb-599415311831.png)

**Comment by optimalisatie** [comment]
@AlphaDE that issue is unrelated to this issue.

`.meta` files are never linked in the HTML, how could Google know about the files and even consider indexing them?

**Comment by Venomz3** [comment]
Yeah I know that there is no sleep mode.  I just used analogy to describe where to compare it. 

Now I got response from server stuff. Server uses PHP opcache  as server addon.  My wordpress used w3 total cache and also WPO plugins.  They think that problem is that there is 3 kind of caches used.  I` would happily remove w3 total cache if too many cache problems is problem and also if can fully replace that plugins functionality with WPO plugins.  

What you think about that ? Could that be the problem ? Sounds logical surely..

**Comment by AlphaDE** [comment]
From another user perspective... I use only the O10N Filecache. Multiple caches usually create overhead but do not improve performance. Go back to one cache and test your site speed at Google or other speed metering tools.

**Comment by Venomz3** [comment]
Yeah I know. But I did not know server caching before they told me that they php opcache.  Most of time I get perfect speed with speed metering tools but only sometimes I got result bad results. Maybe multiple caches are doing inconsistencies.  And  I also have to ask server stuff to disable opcache to my site if needed. I`ll wait also for optimalisatie to know he`s thinking about new information.

**Comment by AlphaDE** [comment]
Why disable opcache? O10N Filecache makes use of opcache.

**Comment by Venomz3** [comment]
I did not know that. Some of stuff is new to me.

**Comment by AlphaDE** [comment]
Don't worry, I learned here also a lot... install O10N Filecache and have a look at the setup. I disabled once opcache because it created problems but found some options to prevent that recently so I reenabled it and use it. The speed boost was significant.

**Comment by Venomz3** [comment]
Thanks AlphaDE, I`ll give it a try a bit later.

**Comment by optimalisatie** [comment]
PHP Opcache is a different type of cache. It is used to make PHP faster and it's very important.

The usage of PHP Opcache by the file cache plugin is a unique innovation in which PHP Opcache is used in a smart way to improve file cache performance, but it's not what PHP Opcache is intended for.

The following article provides more information:

https://support.cloud.engineyard.com/hc/en-us/articles/205411888-PHP-Performance-I-Everything-You-Need-to-Know-About-OpCode-Caches

In regards to cache layers, we would advise to consider using Google Cloud CDN as a second cache layer for SEO purposes. It's more difficult to use as the cache can't be cleared automatically / from within the WordPress plugin. You would need to login to Google Cloud to clear the cache using a optional wild card, and it will cost $0.005 per cache invalidation.

When using Google Cloud CDN as a cache layer, the website will load from Google's network. It will have the lowest latency for all individual users and the benefit for SEO is that Google knows that the performance of the CDN is capable of handling 1 million visitors in 1 minute, so that when search traffic for a specific term spikes it will logically more likely select a Google CDN based website for a #1 position / promotion on mobile search results to satisfy their users.

![image](https://user-images.githubusercontent.com/8843669/41896455-efb6bc74-7924-11e8-9864-adad8891ac25.png)

Our [Google CDN Cache plugin](https://github.com/o10n-x/wordpress-google-cdn-page-cache) is a first prototype but the technical concept simply works as can be seen on www.e-scooter.co. 

You can use the free micro instance to setup a simple Nginx proxy that routes traffic from the CDN to your existing server. The Google CDN plugin then controls the cache using HTTP headers.

https://github.com/o10n-x/wordpress-google-cdn-page-cache

An extra advantage from Google's CDN is that a website will be fast from any location in the world:

![image](https://user-images.githubusercontent.com/8843669/41896737-ae6e02e4-7925-11e8-8604-1ec3b38657f1.png)

And you can use a button on the website:

![image](https://user-images.githubusercontent.com/8843669/41896838-f7ea80c8-7925-11e8-86cb-7aff9a71e75c.png)

More info: https://cloud.google.com/cdn/

**Comment by Venomz3** [comment]
Yes, I`m been thinking about Google CDN.  I got few things to add site before that.  When I disabled W3 total cache ja added O10N Filecache wp admin seems more faster. And Also Pingdom tools reports 2 sec loading time of my site.  And with closest server it reports less than minute.. I had load times 250ms but those weren`t consistent results as they vary a lot.     


All errors seems to be gone. I had lot of testing with many settings and leaving wp admin idle and then after while going to WPO setting and not a single error. But I want to test with W3 total cache with php opcache setting enable to see if it`s only needed that all performance plugins use php opcache.   So dont`t mark this as resolved yet. 


Also when I disabled W3 total cache the I got 2 errors more with google Pagespeed.  

1.Leverage browser caching
Setting an expiry date or a maximum age in the HTTP headers for static resources instructs the browser to load previously downloaded resources from local disk rather than over the network.
2. Enable compression
Compressing resources with gzip or deflate can reduce the number of bytes sent over the network.

Is there a way to remove those with WPO plugins so that I do not need W3 total cache anymore ? 

Site note: Seems to me that now when made following changes I can enable more settings with WPO plugins without getting slow server error from Pagespeed.

**Comment by optimalisatie** [comment]
Compression and caching headers are sent by the server. The best performance is achieved when the configuration is added within the webserver. On Apache servers it's possible to use `.htaccess` but `vhost.conf` may provide (slight) better performance for production servers. On Nginx servers you would need to edit the Nginx server config or the vhost management `nginx.conf` include (when provided by server management software such as Plesk or Cpanel). On regular Unix servers you may find the config files in `/etc/nginx/`.

https://www.linode.com/docs/web-servers/nginx/how-to-configure-nginx/

W3 Total Cache simply writes to the `.htaccess` file so it appears that you are using Apache.

https://gist.github.com/Vheissu/5000708 (W3 Total Cache .htaccess config)

We provide two plugins that enable to tune `.htaccess` or `nginx.conf`. They merely provide a graphical editor that enables efficient optimization.

### .htaccess editor
https://github.com/o10n-x/wordpress-htaccess-editor 

![htaccess-editor](https://user-images.githubusercontent.com/8843669/41925912-27cfacc6-796e-11e8-91e2-c26041ad75b5.png)

### nginx.conf editor
https://github.com/o10n-x/wordpress-nginx-config-editor

![nginx-config-editor](https://user-images.githubusercontent.com/8843669/41925913-27f4a6b6-796e-11e8-8e71-53c14e831337.png)

In regards to settings, regrettably that may vary per server OS / environment. It may be best to find a matching online tutorial for your server, e.g. "Cent OS apache server optimization". 

The following article may provide valuable information for Nginx optimization:

https://gist.github.com/denji/8359866

**Comment by Venomz3** [comment]
Thank you for that information. When I removed W3 total cache and I get once error:

Failed to store cache file /wp-content/cache/o10n/css/src/2f/25/2c/40ad1e46563958db436bca6b74.css
File is not writeable: /wp-content/cache/o10n/css/src/2f/25/2c/40ad1e46563958db436bca6b74.css
Failed to store cache file /wpcontent/cache/o10n/css/proxy/12/5c/16/f039cd8194c59b4ca06d17fbeb.css
File is not writeable: /wp-content/cache/o10n/css/proxy/12/5c/16/f039cd8194c59b4ca06d17fbeb.css


Then I added W3 Total Cache .htaccess config to my .htaccess. Added it with bulletproof securitys .htacess editor.  Then I get slow server form Pagespeed. And also I get another  time those errors.  I quess that errors are removed if I remove new code from .htaccess . 
I`m going to remove those lines from .htaccess and see what happens. 


Just I thought that maybe I should have empty caches before remove W3 total cache plugin… 
 Maybe I should reset PHP opcache with o10n File Page Cache plugin… 


So 2 times I got error: When removing W3 total cache and clicking to o10n  menu and when I added
those lines to .htacess

I have to admit that I`m getting frustrated as I already thought that this issue is resolved. But I`m still investigating what steps solve or throw error.

**Comment by Venomz3** [comment]
Update: Bulletproof security (Bps) plugin removes  o10n top menu when clicking any of bps menu items on left menu and when clicking some other item restores menu. I`ll disable bps tomorrow and test some more.  bps wizard should whitelist  o10n plugins automatically when running wizard.  I`ll continue tomorrow with testing.

**Comment by optimalisatie** [comment]
Hi @Venomz3,

In regards to your latest message, gzip is CPU intensive. It may be that access to CPU is restricted and just a slight delay in response has a significant impact on website speed.

**Comment by Venomz3** [comment]
Hi optimalizatie,
Ok. . I cannot get original error anymore with newest plugin release. Instead I get multiple  "failed to proxy" messages when I login wp admin and click 100 optimization. 
 I did not know when to put it because it`s not specific to some plugin as it`s in main page where you can select different WPO plugins.  If I click second time then there is no error messages.  

And if I take hour about then I get those error messages when I click that 100 optimization link.  
![failed to proxy](https://user-images.githubusercontent.com/39744338/49730896-6083d280-fc82-11e8-80c2-bfdc67e257d9.PNG)

**Comment by optimalisatie** [comment]
Hi @Venomz3,

The error notices indicate that the (PHP based) script proxy wasn't able to reach the specified URLs. The `Status: 404`  code indicates that the specified URLs are returning a 404 HTTP status header. That may be a clue.

The script proxy provided by the plugin uses `wp_remote_get` and `wp_remote_head`, default WordPress request methods that are automatically optimized to use Curl or for example special outgoing proxy settings for regular WordPress usage (e.g. updating plugins). So the proxy should be very stable and is not likely to be a cause of an error.

The proxy provided by the plugin is optimized. For example, when possible, updates are processed in the background via a PHP shutdown method so that they do not impact the page load performance and a conditional request based on `Last-Modified` or `Etag` is used to check for a `304 - Not modified` status header so that assets are downloaded only when they are updated, making the solution efficient for both server performance (for every single visitor / at all times) and data-transfer saving.

Do you have an idea why the URLs specified in the errors could return 404 seen from PHP's (server environment) perspective?

**Comment by Venomz3** [comment]
Hi, 
Now I got time to get into this problem. Those 404 files were my site`s Facebook window JavaScript files. 
I disabled proxy capture in js delivery optimization and error seems to be gone.  Also that had another effect also. Loading time of my site increased significally.  I wondered why it had gone slow..  That was because it was trying to found those files.  I got zero error message from WPO plugins now.
